### ISSUE TITLE
Automation RM cmdlets don't adhere to erroractionpreference

### ISSUE THREAD

**Issue by trondhindenes** [Azure/azure-powershell#1135]
Repro:
`Get-AzureRmAutomationRunbook -Name "DoesNotExist" -ErrorAction SilentlyContinue`

Result: Throws error
Expected result: supresses error

I thought we were passed these bugs...

**Comment by devigned** [comment]
@jodoglevy can we get this fixed up prior to 1.0 release.

**Comment by jodoglevy** [comment]
@trondhindenes this is specific to automation cmdlets? I don't think we're doing anything specific with error action

**Comment by trondhindenes** [comment]
Hard to say - The automation cmdlets are the only ones I've been using. But you're definetely doing erroraction wrong in the current ver (and not just in the cmdlet I used in my example).

**Comment by jodoglevy** [comment]
I just tried with Get-AzureRmVM, Same issue. This is not specific to Automation cmdlets. It's a problem with all of Azure PS 1.0 preview.

**Comment by devigned** [comment]
@singhkay looks like this is happening in VM cmdlets too.

**Comment by singhkay** [comment]
@hyonholee Can you check the VM cmdlets in the preview module?

**Comment by markcowl** [comment]
@singhkay @hyonholee @trondhindenes @jodoglevy   This is absolutely expected.  Terminating errors should result in failures regardless of ErrorActionPreference.  Only non-terminating ewrrors should be controlled this way, see https://msdn.microsoft.com/en-us/library/ms714414(v=vs.85).aspx

**Comment by trondhindenes** [comment]
Yup, tgat makes sense. Tho I don't understand the rationale for throwing terminating errors.

**Comment by jodoglevy** [comment]
It appears these are exceptions, although I don't think they should be:

`try {
    Get-AzureRMVm -ResourceGroupName "dsfd" -Name "fdsfds"
}
catch
{
    "EXCEPTION!!"
}`

results in "EXCEPTION!!" but

`try {
    Get-AzureVM -Name "sdds" -servicename "fdssfd"
}
catch
{
    "EXCEPTION!!"
}`

does not.

Same with automation cmdlets:

`try {
    Get-AzureRmAutomationRunbook -ResourceGroupName "dfds" -AutomationAccountName "ddd" -Name "sdsd"
}
catch
{
    "EXCEPTION!!"
}`

results in "EXCEPTION!!" but

`try {
    Get-AzureAutomationRunbook -AutomationAccountName "ddd" -Name "sdd"
}
catch
{
    "EXCEPTION!!"
}`

does not

@safeermohammed any ideas?

**Comment by trondhindenes** [comment]
Terminating/Non-terminating errors need to be absolutely predictable - otherwise it is impossible to effectively build solutions using the module - especially headless scripts where we need to be super-picky about error-handling. Please get this fixed.

**Comment by jodoglevy** [comment]
@safeermohammed did we change anything with regards to error handling from our ASM to our ARM cmdlets?

**Comment by safeermohammed** [comment]
@jodoglevy . None I am aware of. @markcowl Can you please take a look at this?

**Comment by trondhindenes** [comment]
There's definetively differences. For example (all against non-existing things):
get-azurevm -Name "smokk" -ServiceName "kvekk": returns warning, but no error
Get-AzureRmVM -ResourceGroupName "kokk" -Name "krokk": returns a TERMINATING error, which is (excuse my french) bollocks. Please read up on the PowerShell design guidelines:

"For a non-terminating error that does not stop operation on the next record that is coming from the pipeline (for example, a record produced by a different process), the cmdlet must call the WriteError method"

given a pip or array of objects, a single not-found vm does not stop PowerShell from processing the next record, and therefore a terminating error should never be thrown. Terminating errors should (IMHO) only be used in cases where the cmdlet loses connection to the underlying rest api or similar. Look at how most (Microsoft-supplied) modules are written, and you will see that they VERY rarely throw terminating errors during "regular operations". Please get this sorted, it is a major bug.

**Comment by devigned** [comment]
@mayurid and @markcowl this sounds like there was a change in behavior. If so, we need to fix this.

**Comment by trondhindenes** [comment]
Looking at some of the modules, I can't see any unit tests covering error schenarios, so no wonder this wasn't picked up. I would advise you to add integration tests covering failure scenarios to make sure cmdlets output the right error type.

**Comment by markcowl** [comment]
@trondhindenes @devigned I think we have let the perfect be the enemy of the good in this case.  It is technically true that the decision whether an error is terminating or non-terminating should be made on a case-by-case basis, but most errors that occur in ProcessRecord will be non-terminating.  We will update the cmdlet advice to designate terminating errors using ThrowTerminatingError and catch other execptions thrown during ProcessRecord and use WriteError() to make them non-terminating

**Comment by jodoglevy** [comment]
Awesome, thanks for looking at this @markcowl.

**Comment by trondhindenes** [comment]
Thanks @markcowl looking forward to this!

**Comment by slapointe** [comment]
Cain't wait to see this fixed as it cause a lot of problems porting our existing scripts.

**Comment by graemechristie** [comment]
Still no resolution on this ? This has broken pretty much every script in out deployment pipeline. I don't see how calling Get-AzureRmWebApp on a web app that doesn't exist should be considered a terminating error.

**Comment by devigned** [comment]
This should be fixed in the upcoming December release.

**Comment by markcowl** [comment]
@trondhindenes @graemechristie @slapointe Since this is also a breaking change, the plan is to lump this change together with the ShouldProcess change and make the change across all the ARM cmdlets. at or just after the //BUILD timeframe.

**Comment by markcowl** [comment]
### Root Cause
Common cmdlet does not place ty/catch around cmdlet execution and write exceptions as ErrorRecords, and ProcessRecord is simply throwing an exception.  This also prevents gatrhering good data about exceptions in telemetry.

### Proposed Solution
- Wrap ProcessRecord execution in base cmdlet, and write error records with caught exception details.  terminating errors in ProcessRecord should call ThrowTerminatingError directly.
-  Add mock tests for terminating and non-terminating process record errors to ensure the correct CommandRuntiem methods are called appropriately in case of errors.
- Similarly, wrap BeginProcessing/Endprocessing in the same fashion and call ThrowTerminatingError instead of writing error record
- Ensure no BeginProcessing errors in Set-AzureRMContext / Get-AzureRmContext

**Comment by mgreenegit** [comment]
New-AzureRMResourceGroupDeployment does not exit in error if one resource within a deployment template fails?  Without ending in error, CI platforms do not know whether the deployment was successful.

**Comment by peterincontext** [comment]
@mgreenegit, agreed, it's not ideal in its current implementation.  However, instead of relying on the actual exit, it's possible to check the ProvisioningState, which does change to **Failed** if any of the sub-deployments in the template fail.

    $deploy = New-AzureRmResourceGroupDeployment @deployParameters -Name $deploymentLabel -Verbose
    
    if($deploy.ProvisioningState -eq "Failed")
    {
    Write-Output "ARMTemplateDeploy Provision Failed."
    Write-Output $deploy
    exit 1
    }

Works great to properly fail in my CI platform.

**Comment by jonlesada** [comment]
Same issue here... with Get-AzureRMResourceGroup.  Was using it to validate the entry of a resource group name and just realized that it isn't working.  I have been using many of the Get-* cmdlets to validate data, and it seems that none of them respond to a try/catch block.  Unfortunately some of the data that I need to validate can't be easily queried so I'll have to loop through the output and check for it myself.

**Comment by abatishchev** [comment]
To get into a try/catch block you need to add parameter `-ErrorAction Stop` to the call each of the Get-* cmdlets. However, `$_` won't have any meaningful information, all fields will be empty.

**Comment by mikewo** [comment]
I've hit this same issue with Get-AzureRMPolicy definition, it doesn't not adhere to the SilentlyContinue on the error action preference if it doesn't find the policy; whereas Get-AzureRmManagementGroup does. Is there any way that something can be put in place so that every cmdlet is tested for this behavior automatically when it is created? The cmdlets as they stand now are very inconsistent in how they handle this.
